plugins/org.eclipse.xtext.util/src/org/eclipse/xtext/util/OnChangeEvictingCache.java

```diff
@@ -109,11 +110,19 @@ protected void cacheHit(CacheAdapter adapter) {
 	 * @return the cache adapter for the given resource. Never <code>null</code>.
 	 */
 	public CacheAdapter getOrCreate(Resource resource) {
-		CacheAdapter adapter = (CacheAdapter) EcoreUtil.getAdapter(resource.eAdapters(), CacheAdapter.class);
-		if (adapter == null) {
-			adapter = new CacheAdapter();
-			resource.eAdapters().add(adapter);
-			adapter.setResource(resource);
+		// ask for the list of adapters twice in order to avoid the usage of a
+		//  potentially stale list because of a race of several involved threads
+		resource.eAdapters();
```

Comment:
This does not prevent a race. Consider two thread T1 and T2:
T1 enters this method and calls resource.eAdapter and is suspended before eAdapters returns and the field of the resource is set.
T2 enters this method, calls eAdapters, creates a new list, calls eAdapter again and obtains the same list and syncs on that one.
T1 is now overriding the adapter list of the resource and syncs on a different instance.